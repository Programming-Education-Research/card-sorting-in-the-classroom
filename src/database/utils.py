from collections.abc import Generator
from contextlib import contextmanager
from sqlite3 import Cursor, Row, Connection


def dict_factory(cur: Cursor, row: Row):
    fields = [column[0] for column in cur.description]
    return {key: value for key, value in zip(fields, row)}


@contextmanager
def cursor(con: Connection) -> Generator[Cursor]:
    cur = con.cursor()
    try:
        cur.row_factory = dict_factory
        yield cur
    finally:
        cur.close()
