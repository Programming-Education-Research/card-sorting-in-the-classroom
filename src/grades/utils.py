from grades.grade import Grade


def from_raw(data: dict[str, str | float]):
    return Grade(
        username=data["username"],
        grade=calculate_average(data),
        sat_exam="Exam" in data,
    )


def calculate_average(data):
    marks = [v for k, v in data.items() if k.startswith("Test") or k == "Exam"]
    return sum(marks) / len(marks) if marks else 0
