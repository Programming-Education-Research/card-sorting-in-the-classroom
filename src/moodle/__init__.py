from . import grades
from . import responses
from . import xml
