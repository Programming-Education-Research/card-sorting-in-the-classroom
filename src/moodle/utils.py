from collections import defaultdict


def join_by(key, *args):
    key_to_result = defaultdict(dict)
    for col in args:
        for item in col:
            key_to_result[item[key]] |= item
    return list(key_to_result.values())