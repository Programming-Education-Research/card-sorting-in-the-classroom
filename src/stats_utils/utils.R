library(tidyverse)
library(DBI)
library(knitr)

question_ordering <- c(
    "Math Expressions",
    "Boolean Reasoning",
    "Dicts",
    "Dicts With Errors",
    "Functional Equivalence",
    "Identify Exceptions",
    "Test Case Match"
)

relabel_courses <- function(course) ordered(
    case_match(course, "CS101" ~ "Compsci 101", "CS130" ~ "Compsci 130"),
    levels = c("Compsci 101", "Compsci 130")
)

relabel_semester <- function(semester) ordered(
    case_match(semester, "SS2025" ~ "Summer Semester", "S12025" ~ "Semester One"),
    levels = c("Summer Semester", "Semester One")
)

relabel_exam <- function(exam) ordered(
    case_match(exam, 1 ~ "Sat Exam", 0 ~ "Did Not Sit"),
    levels = c("Sat Exam", "Did Not Sit")
)

relabel_questions <- function(questions) ordered(
    questions,
    levels = question_ordering
)

t <- function(table) {
    if (isTRUE(getOption('knitr.in.progress'))) {
        return(knitr::kable(table))
    } else {
        return(table)
    }
}


fetch <- function(query) {
    mydb <- dbConnect(RSQLite::SQLite(), "data/database.db")
    raw_data <- dbGetQuery(mydb, query)
    dbDisconnect(mydb)
    return(raw_data)
}


custom_theme <- function(...) (
    theme_minimal()
    + theme(
        plot.title = element_text(face = "bold"),
        axis.text = element_text(colour = "black", face = "bold", size = 10),
        axis.title = element_text(colour = "black", face = "bold", size = 11),
        panel.grid = element_line(
            colour = "black", linetype = "dotted", linewidth = 0.3),
        panel.border = element_rect(linetype = "solid", fill = NA),
        strip.text = element_text(colour = "black", face = "bold", size = 12),
        plot.caption.position = "plot",
        plot.background = element_rect(fill = "white", colour = "white"),
    )
    + theme(...)
)

is_outlier <- function(values) {
    return(values > quantile(values, 0.75) + 1.5 * IQR(values, na.rm = TRUE))
}
