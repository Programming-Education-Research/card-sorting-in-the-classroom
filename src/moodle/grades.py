import csv
from collections import defaultdict
from datetime import datetime
from typing import IO

from .utils import join_by


def load(names_to_paths):
    tests = []
    for test_name, filename in names_to_paths.items():
        with open(filename, "r", encoding="utf-8-sig") as f:
            result = load_attempts(f, test_name)
            tests.append(result)
    return join_by("username", *tests)


def load_attempts(grades: IO, name: str):
    attempts = first_attempts(grades)
    attempts = [select_columns(attempt, name) for attempt in attempts]
    return attempts


def select_columns(attempt, name):
    grade_col = next(k for k in attempt.keys() if k.startswith("Grade/"))
    grade = float(attempt[grade_col])
    out_of = float(grade_col.split("/")[1])
    return {
        "username": attempt["Username"],
        name: grade / out_of
    }


def first_attempts(grades: IO):
    reader = csv.DictReader(grades)
    username_to_attempts = defaultdict(list)
    for row in reader:
        if row["Last name"] != "Overall average":
            username_to_attempts[row["Username"]].append(row)
    return [
        # attempts
        min(attempts, key=lambda a: parse_time(a["Started on"]))
        for attempts in username_to_attempts.values()
    ]


def parse_time(timestamp: str) -> datetime:
    return datetime.strptime(timestamp, "%d %B %Y  %I:%M %p")
