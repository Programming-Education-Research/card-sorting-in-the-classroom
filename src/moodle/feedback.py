from typing import IO

from moodle import responses


def load(f: IO, rating_num: int, comment_num: int):
    data = responses.load(f)
    ratings = {u: a[-1] if a else None for u, a in data[rating_num].items()}
    comments = {u: a[-1] if a else None for u, a in data[comment_num].items()}
    return join(ratings, comments)


def join(ratings, comments):
    return {
        username: {
            "rating": ratings.get(username),
            "comment": comments.get(username)
        }
        for username in set(ratings).union(comments)
    }
