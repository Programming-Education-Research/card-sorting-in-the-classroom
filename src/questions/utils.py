from moodle.xml import RawQuestion
from questions.cardsort import Cardsort
from questions.question import Question
from questions.refute import Refute
from questions.reverse_trace import ReverseTrace


def from_raw(raw: RawQuestion) -> Question:
    match raw["type"]:
        case "python_refute":
            return Refute.from_raw(raw)
        # I have no idea why, but the production Moodle instance suffixes a 1
        case "python3_html_gapfiller-1":
            return ReverseTrace.from_raw(raw)
        case "cardsort":
            return Cardsort.from_raw(raw)
        case _:
            raise RuntimeError(f"Unrecognised question type: `{raw['type']}`")
