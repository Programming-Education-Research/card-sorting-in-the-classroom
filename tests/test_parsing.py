from pprint import pformat

from approvaltests import verify

import moodle
from questions.utils import from_raw
from utils import test


@test
def refute_and_reverse_trace_questions_can_be_parsed():
    with open("./data/questions.xml", "rb") as f:
        questions = [from_raw(q) for q in moodle.xml.parse(f)]
    verify(pformat(questions, width=240))
