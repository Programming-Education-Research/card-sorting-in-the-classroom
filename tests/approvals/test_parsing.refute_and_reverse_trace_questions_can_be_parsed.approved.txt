[Refute(name='refute: bring umbrella',
        prompt='Your classmate wrote the following function that takes three boolean \r\n'
               'variables to help them decide whether they should take an umbrella with \r\n'
               'them:\n'
               '\n'
               'def take_umbrella(is_cloudy, is_raining, going_out):\r\n'
               '    if is_cloudy or is_raining and going_out:\r\n'
               '        return "Take an umbrella"\r\n'
               '    else:\r\n'
               '        return "Don\'t take an umbrella"\n'
               '\n'
               '\n'
               'Your classmate will bring an umbrella with them if:\n'
               '\n'
               'It is cloudy, or it is raining\n'
               'But only if they are going out for the day\n'
               '\n'
               'In all other cases, they will not take an umbrella with them.\n'
               '\n'
               'There is a bug in their code – write a test case that reveals the bug. Remember to include quote marks around any strings.',
        preload='Calling `take_umbrella({[24]})`\nShould return: {[25]}\nBut the above function actually returns: {[25]}',
        test='def buggy(is_cloudy, is_raining, going_out):\n'
             '    if is_cloudy or is_raining and going_out:\n'
             '        return "Take an umbrella"\n'
             '    else:\n'
             '        return "Don\'t take an umbrella"\n'
             '\n'
             'def correct(is_cloudy, is_raining, going_out):\n'
             '    if (is_cloudy or is_raining) and going_out:\n'
             '        return "Take an umbrella"\n'
             '    else:\n'
             '        return "Don\'t take an umbrella"\n'
             '\n'
             'assert correct({[given]}).lower() == {[then]}.lower()\n'
             'assert buggy({[given]}).lower() == {[but]}.lower()'),
 Refute(name='refute: no pairs over 5',
        prompt='A classmate has written the following function:\n'
               '\n'
               'def no_large_pairs(data):\r\n'
               '    for i in range(len(data)):\r\n'
               '        if data[i - 1] + data[i] > 5:\r\n'
               '            return False\r\n'
               '    return True\n'
               '\n'
               'They claim the function returns True if there is no adjacent pair of numbers that sum to a value greater than five, and False otherwise.There is a bug in their code – write a test case that reveals the bug.',
        preload='Calling no_large_pairs({[16]})\nShould return: {[8]}\nBut the above function actually returns: {[8]}',
        test='def buggy(data):\n'
             '    for i in range(len(data)):\n'
             '        if data[i - 1] + data[i] > 5:\n'
             '            return False\n'
             '    return True\n'
             '\n'
             'def correct(data):\n'
             '    for i in range(len(data) - 1):\n'
             '        if data[i] + data[i + 1] > 5:\n'
             '            return False\n'
             '    return True\n'
             '\n'
             'assert correct({[given]}) == {[then]}\n'
             'assert buggy({[given]}) == {[but]}'),
 ReverseTrace(name='Reverse Trace: flip evens in list',
              prompt='Complete the following code to get the expected output in the example box below.',
              preload='def munge(data):\n    for i in range(len(data) // 2):\n        if i % 2 == 0:\n            data[i], data[-i - 1] = data[-i - 1], data[i]\n\n\ndata = {[45]}\nmunge(data)\nprint(data)',
              expect="['a', 'b', 'c', 'd', 'e', 'f', 'g']")]
